### PR TITLE
Fix FastJet cmake setup

### PR DESCRIPTION
**Pull request by matthiasrichter**
FastJet is an optional package for O2 and this needs to be reflected in the
cmake setup
- make FindFastJet not to exit with FATAL_ERROR unless REQUIRED
- fix FindFastJet bahavior for CONFIG/MODULE mode
- mark FastJet as optional for O2

### CODE REVIEW

**Comment by davidrohr** [comment]
@aphecetche : I think this was introduced during the CMake migration. Could you have a look?

**Comment by aphecetche** [comment]
@davidrohr no, it was introduced more recently by @qgp in https://github.com/AliceO2Group/AliceO2/pull/3041, so I'll let him comment on this PR, but from my side, it looks good.

File: dependencies/FindFastJet.cmake [hunk 1]
```diff
@@ -68,28 +68,35 @@ if(${PKGNAME}_CONFIG)
...
+      else()
+        get_filename_component(dir_gmp ${lib_gmp} DIRECTORY)
+        list(APPEND ${PKGNAME}_LIB_DIRS ${dir_gmp})
       endif()
-      get_filename_component(dir_gmp ${lib_gmp} DIRECTORY)
-      list(APPEND ${PKGNAME}_LIB_DIRS ${dir_gmp})
     endif()
   endif(${PKGNAME}_CONFIGLIBS)
-endif(${PKGNAME}_CONFIG)
+else(${PKGNAME}_CONFIG)
```
Comment (qgp) [inline dependencies/FindFastJet.cmake:92]: The following code should also be evaluated in order to properly said the variables (even if NOT_FOUND).

File: dependencies/FindFastJet.cmake [hunk 2]
```diff
@@ -68,28 +68,35 @@ if(${PKGNAME}_CONFIG)
     if ("CGAL" IN_LIST ${PKGNAME}_LIBS)
       find_library(lib_cgal NAMES "CGAL" PATHS $ENV{CGAL_ROOT}/lib NO_DEFAULT_PATH)
       if (NOT lib_cgal)
-        message(FATAL_ERROR "CGAL not found in $ENV{CGAL_ROOT}/lib")
+        if(FastJet_FIND_REQUIRED)
```
Comment (qgp) [inline dependencies/FindFastJet.cmake:71]: This part is only reached if `fastjet-config` is found, and in this case there is a problem if it reports the use of CGAL but we cannot find the library.

File: dependencies/FindFastJet.cmake [hunk 3]
```diff
@@ -68,28 +68,35 @@ if(${PKGNAME}_CONFIG)
...
       endif()
-      get_filename_component(dir_cgal ${lib_cgal} DIRECTORY)
-      list(APPEND ${PKGNAME}_LIB_DIRS ${dir_cgal})
     endif()
 
     if ("gmp" IN_LIST ${PKGNAME}_LIBS)
       find_library(lib_gmp NAMES "gmp" PATHS $ENV{GMP_ROOT}/lib NO_DEFAULT_PATH)
       if (NOT lib_gmp)
-        message(FATAL_ERROR "GMP not found in $ENV{GMP_ROOT}/lib")
+        if(FastJet_FIND_REQUIRED)
```
Comment (qgp) [inline dependencies/FindFastJet.cmake:83]: This part is only reached if fastjet-config is found, and in this case there is a problem if it reports the use of GMP but we cannot find the library.

**Comment by qgp** [comment]
@matthiasrichter I am not sure I understand your logic here. If FastJet reports that it is linked agains CGAL/GMP (through `fastjet-config --libs`) and we don't find the respective library, that is an issue. If FastJet as such is not found, we never reach that point. Am I missing something?

Agreed for the property to mark it as optional

File: dependencies/FindFastJet.cmake (hunk 2 above)
Comment (aphecetche) [reply dependencies/FindFastJet.cmake:71]: I guess @matthiasrichter goal here was to prevent a plain crash of the CMake configure stage in cases where find_package(FastJet) is called without the REQUIRED argument ? The fact that CGAL is not found (where it should be) is indeed an issue, but should only be fatal one if FastJet is required.

Comment (qgp) [reply dependencies/FindFastJet.cmake:71]: @aphecetche I disagree, finding fastjet exposing linkage against CGAL but not finding CGAL itself would result in an inconsistent usage. This should not happen and, for me, such an inconsistency (package exposing another library through `<pkg>-config`, the latter one not found) warrants a failure of the configuration. 

If we deliberately decide to handle this gracefully (even though I don't see a good reason to do so), we would have to invalidate the finding of FastJet itself and treat also FastJet as not found. 

@matthiasrichter In order to better understand what you were after here, did you actually observe a failure related to this? If so, under what conditions?

Comment (aphecetche) [reply dependencies/FindFastJet.cmake:71]: > If we deliberately decide to handle this gracefully (even though I don't see a good reason to do so), we would have to invalidate the finding of FastJet itself and treat also FastJet as not found.

Sure. If anything goes wrong in the FindFastJet.cmake, then FastJet must be declared NOT FOUND.

**Comment by matthiasrichter** [comment]
@qgp, in general  I do not expect a FATAL_ERROR if I look for an optional package, and cmake can not resolve all dependencies (meaning the optional package is not available). The fatal error simply crashes cmake configuration. It is also suppressing the feature summary we are using at the end of cmake configuration.

I want to set up a standalone project using O2, and in this setup optional packages should be silently skipped.

Why does FindFastjet need to check for those two dependencies, is this not done in the config file already?

Sure, we have to make sure that `FastJet_FOUND` is not set. This is actually the case in my setup, after executing `fastjet-config`, the variable is not set. Maybe the config file already checks for the other dependencies? But to be sure this should be unset if the dependencies are not fulfilled. We can also print a warning unless FastJet_FIND_QUIET is set.

In summary, something is incomplete in the way we have the check right now, 
- it can break cmake configuration for an optional package
- maybe fastjet config file is incomplete, because additional checks are necessary in the find macro

**Comment by qgp** [comment]
@matthiasrichter We need the `FindFastJet.cmake` because FastJet does not provide a `FastJetConfig.cmake`. I am fine with making FastJet optional, but if FastJet is found it should be expected to be properly set up (i.e. required libraries are available). Maybe quicker to resolve in a quick Skype call?

**Comment by matthiasrichter** [comment]
@qgp I have updated the PR, some lines only have inserted blanks because of an additional if-clause, use diff with whitespace suppression, to see relevant changes
- the extra checks are now executed only if FastJet_FOUND is set by the config command
- removed the else branch, so `find_package_handle_standard_args` is always executed

I though the latter one is complimentary to the CONFIG mode and should not be executed when config command is found.

A find macro should only throw fatal errors if the package is required, I think the current version implements this appropriately, also taking your comments into account.

File: dependencies/FindFastJet.cmake [hunk 4]
```diff
@@ -46,43 +46,47 @@ if(${PKGNAME}_CONFIG)
...
-        string(SUBSTRING ${libdir} 2 -1 libdir)
-        list(APPEND ${PKGNAME}_LIB_DIRS ${libdir})
-    endforeach()
-
-    # append directories of exposed libraries
-    if ("CGAL" IN_LIST ${PKGNAME}_LIBS)
-      find_library(lib_cgal NAMES "CGAL" PATHS $ENV{CGAL_ROOT}/lib NO_DEFAULT_PATH)
-      if (NOT lib_cgal)
-        message(FATAL_ERROR "CGAL not found in $ENV{CGAL_ROOT}/lib")
+  if (${PKGNAME}_FOUND})
```
Comment (qgp) [inline dependencies/FindFastJet.cmake:49]: I fail to see how this is supposed to work, `${PKGNAME}_FOUND` will only be set by the macro in line 93 (in your version). What do I miss?

In fact, I don't see the need for any protection here.

Comment (matthiasrichter) [reply dependencies/FindFastJet.cmake:49]: you need to hide whitespace changes, this protection is actually at a different place as you might think when looking at the full diff. `${PKGNAME}_FOUND` is set by the configuration command, and only i the configuration command finds the package, the additional checks should be executed.

Comment (matthiasrichter) [inline dependencies/FindFastJet.cmake:49]: @qgp let me check if github produces the whitespace-supressed diff on this comment.

Comment (matthiasrichter) [reply dependencies/FindFastJet.cmake:49]: no, did not work

Comment (qgp) [reply dependencies/FindFastJet.cmake:49]: @matthiasrichter Which "configuration command" are you talking about? Lacking a FastJetConfig.cmake, we fall back to `fastjet-config`, but this does not set any cmake variables.

PS: Yes, I looked at the full file (and with whitespace changes hidden).

Comment (matthiasrichter) [reply dependencies/FindFastJet.cmake:49]: Correct, I'm on the wrong path.

**Comment by matthiasrichter** [comment]
@qgp So what is your proposal to solve the problem, the FATAL_ERROR destroys the possibility to have FastJet as an optional package, if the package config is not consistent. Go back to my original version with FATAL_ERROR only if FIND_REQUIRED. And a warning if not FIND_QUIET?

**Comment by qgp** [comment]
> @qgp So what is your proposal to solve the problem, the FATAL_ERROR destroys the possibility to have FastJet as an optional package, if the package config is not consistent. Go back to my original version with FATAL_ERROR only if FIND_REQUIRED. And a warning if not FIND_QUIET?

@matthiasrichter No, if we just make this a warning, we would still use FastJet but with a broken setup. While I still don't understand why supporting a broken FastJet installation is so important to you, I would suggest to set define an (advanced) variable, e.g. `${PKGNAME}_VALID` when `fastjet-config` is found, unset it in case GMP/CGAL is used but not found (there should still be an error message to make obvious why FastJet is not picked up), and add this variable to the required variables for `find_package_handle_standard_args`. Does this make sense to you?

**Comment by matthiasrichter** [comment]
So we need to skip `find_package_handle_standard_args`, how about the updated version?

cmake message only provides FATAL_ERROR and SEND_ERROR, both abort, though the latter runs the configuration through. That means WARNING is the message type we can use.

Why did I end up in FindFastJet? Because I failed to use the O2 installation in a stand-alone package. My users story is:
I'm not interested in a broken FastJet, in fact it's an optional package and thus should simply behave like a silent spectator. It's important for me to have a working cmake setup. I want `FastJet_FOUND` not to be set if anything in the find process for this optional package is incomplete. Since it's optional I don't want to struggle with it. cmake feature summary should tell me that the optional package FastJet is not found. And if I'm interested in more I can check the messages higher up. In principle I am interested, and that's the second reason why I'm here. but people I'm preparing the sandbox for are not.

**Comment by qgp** [comment]
@matthiasrichter Thanks for the explanation (even though I still don't know why the fastjet installation is broken instead of simply fastjet not being present, but doesn't matter).

And, no, we should not just skip `find_package_handle_standard_args` but make it fail with one of the variables listed as required undefined in order to make it properly set FastJet as `NOT_FOUND`.

**Comment by sawenzel** [comment]
What is the status here? Is this still being worked on?

**Comment by matthiasrichter** [comment]
I have no time to work on that, can reopen if I need to work on that again.
